azurerm/resource_arm_iothub_resource_provider.go

```diff
@@ -0,0 +1,264 @@
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/response"
+)
+
+func resourceArmIotHub() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmIotHubCreateAndUpdate,
+		Read:   resourceArmIotHubRead,
+		Update: resourceArmIotHubCreateAndUpdate,
+		Delete: resourceArmIotHubDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"tags":     tagsSchema(),
+			"location": locationSchema(),
+
+			"resource_group_name": resourceGroupNameSchema(),
+			"sku": {
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:             schema.TypeString,
+							Required:         true,
+							DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+							ValidateFunc: validation.StringInSlice([]string{
+								string(devices.F1),
+								string(devices.S1),
+								string(devices.S2),
+								string(devices.S3),
+							}, true),
+						},
+
+						"tier": {
+							Type:             schema.TypeString,
+							Required:         true,
+							DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+							ValidateFunc: validation.StringInSlice([]string{
+								string(devices.Free),
+								string(devices.Standard),
+							}, true),
+						},
+
+						"capacity": {
+							Type:         schema.TypeInt,
+							Required:     true,
+							ValidateFunc: validation.IntAtLeast(1),
+						},
+					},
+				},
+			},
+
+			"type": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"hostname": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"shared_access_policy": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"key_name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"primary_key": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"secondary_key": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"permissions": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+
+}
+
+func resourceArmIotHubCreateAndUpdate(d *schema.ResourceData, meta interface{}) error {
+
+	armClient := meta.(*ArmClient)
+	iothubClient := armClient.iothubResourceClient
+	ctx := armClient.StopContext
+
+	rg := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+
+	updateIoT := ""
+
+	res, err := iothubClient.CheckNameAvailability(ctx, devices.OperationInputs{
+		Name: &name,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if !*res.NameAvailable {
```

Comment:
will this fail when updating a resource?

Reply:
It updates the resource correctly. I want to add an accTest for this as well.